### DISCUSSION TITLE
AOT compiling with errors messages

### DISCUSSION THREAD

**Discussion by maemigh** [sinclairzx81/typebox#797] [Q&A] [answered]
Is it possible to use the AOT code and check for errors? It doesn't look like TypeCompiler.Code() creates an error function.

**Comment by sinclairzx81** [comment, marked as answer]
@maemigh Hi,

> Is it possible to use the AOT code and check for errors? It doesn't look like TypeCompiler.Code() creates an error function.

Not at this stage. Currently the AOT option will generate a `Check` function that returns a boolean result only. Note that JIT option also only generates the same `Check` (with the Errors obtained through dynamic introspection (the same as `Value.Errors()`)). The current AOT implementation in TypeBox was mostly written to try and accelerate the Check phase in eval restricted environments, using non-eval techniques like the following....

### AOT Dynamic

> The following uses ESM and memory URL sourcing to by-pass direct eval code evaluation. This approach depends on platform features (specifically URL cache), and can be used to obtain high performance checking by generating static AOT code loaded via ESM from an in-memory URL cache. The loaded ESM function is then used with TypeBox's `TypeCheck<T>` infrastructure.

```typescript
import { TypeCompiler, TypeCheck } from 'npm:@sinclair/typebox/compiler'
import { Type, TSchema } from 'npm:@sinclair/typebox'

// (Deno | Browser) Needs ESM + URL.createObjectURL() cache support
export async function DynamicCompile<T extends TSchema>(schema: T, references: TSchema[] = []): Promise<TypeCheck<T>> {
  const code = `export const T = (function () {${TypeCompiler.Code(schema, references)}})()`
  const url = URL.createObjectURL(new Blob([code], { type: 'application/javascript' }))
  return new TypeCheck(schema, references, (await import(url)).T, code)
}

const C = await DynamicCompile(Type.Object({
  x: Type.Number(),
  y: Type.Number(),
  z: Type.Number(),
}))

const R = C.Decode({ x: 1, y: 2, z: 3 })

```
At this stage, the AOT and JIT still depend on dynamic introspection for fail cases (which includes Errors) and as of today, is not part of the compiler code gen infrastructure, but it may be in future (once dynamic API's have stabilized). So, if you are using `.Code()` AOT in isolation (i.e. writing check functions to disk), you will likely be limited to `Check` only.

Hope this helps
S
